ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -804,19 +843,42 @@ void handleBody(InputStream payload)
               MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent, blobIdFactory);
           chunkSize = compositeBlobInfo.getChunkSize();
           totalSize = compositeBlobInfo.getTotalSize();
-          List<StoreKey> keys = compositeBlobInfo.getKeys();
-          chunkIdIterator = keys.listIterator();
-          numChunksTotal = keys.size();
-          dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
-          for (int i = 0; i < dataChunks.length; i++) {
-            dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+          if (hasRange && (chunkSize == CompositeBlobInfo.UNDEFINED_CHUNK_SIZE
+              || totalSize == CompositeBlobInfo.UNDEFINED_TOTAL_SIZE)) {
+            onInvalidRange("Range requests not supported by this metadata content version");
+          } else if (hasRange && !options.getRange().withinTotalSize(totalSize)) {
+            onInvalidRange("Provided getBlob range exceeds the total blob size");
+          } else {
+            List<StoreKey> keys = compositeBlobInfo.getKeys();
+            if (hasRange) {
+              GetBlobRange range = options.getRange();
+              // Get only the chunks within the range.
+              firstChunkOffset = (int) (range.getStartOffset(totalSize) / chunkSize);
+              keys = keys.subList(firstChunkOffset, (int) ((range.getEndOffset(totalSize) / chunkSize) + 1));
+              routerMetrics.onValidRangeRequest(range, totalSize);
```

Comment:
I didn't, other than counting rangeNotSatisfiable errors and range request sizes. I will add these in, though.